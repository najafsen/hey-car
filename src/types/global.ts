import { ReactNode } from "react";

export interface PropsWithChildren<TChildren = ReactNode> {
  children?: TChildren;
}
