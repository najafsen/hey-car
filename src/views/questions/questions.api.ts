import { axiosInstance } from "../../utils/axios";

export const baseUrl = "/questions";

export const getQuestionsList = async () => {
  try {
    const { data } = await axiosInstance.get(baseUrl);
    return data;
  } catch (e) {
    console.warn("Do some error handling!");
  }
};
