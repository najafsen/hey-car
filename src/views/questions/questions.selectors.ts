import { RootState } from "../../app.store";

export const selectQuestionsList = (state: RootState) => state.questions.list;
