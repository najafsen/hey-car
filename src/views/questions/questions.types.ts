export interface Question {
  question: string;
  published_at: string;
  url: string;
  choices: QuestionChoice[];
}

export interface QuestionChoice {
  choice: string;
  url: string;
  votes: number;
}
