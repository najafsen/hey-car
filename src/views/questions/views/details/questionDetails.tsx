import { FunctionComponent } from "react";

export const QuestionDetails: FunctionComponent = () => {
  return <h1>Details</h1>;
};
