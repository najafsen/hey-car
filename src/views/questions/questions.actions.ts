import { createAsyncThunk } from "@reduxjs/toolkit";
import {
  getQuestionById,
  getQuestionsList,
  voteQuestionChoice,
} from "./questions.api";
import { Question, QuestionChoice } from "./questions.types";

export const prefix = "questions";

export const getQuestionsListAction = createAsyncThunk<{
  questions: Question[];
}>(`${prefix}/getList`, async () => {
  const questions = await getQuestionsList();
  return { questions };
});

export const getQuestionByIdAction = createAsyncThunk<
  {
    question: Question;
  },
  string
>(`${prefix}/getById`, async (questionId: string) => {
  const question = await getQuestionById(questionId);
  return { question };
});

export const voteQuestionChoiceAction = createAsyncThunk<
  {
    choice: QuestionChoice;
    questionId: string;
  },
  { questionId: string; choiceId: string }
>(`${prefix}/voteChoice`, async ({ questionId, choiceId }) => {
  const choice = await voteQuestionChoice(questionId, choiceId);
  return { choice, questionId };
});
