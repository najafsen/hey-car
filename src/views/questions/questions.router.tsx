import { Switch, Route } from "react-router-dom";
import { QuestionDetails } from "./views/details/questionDetails";
import { QuestionsList } from "./views/list/questionsList";

export const questionsPageBasePath = "questions";

export const QuestionsRouter = () => (
  <Switch>
    <Route exact path={`/${questionsPageBasePath}`} component={QuestionsList} />
    <Route path={`/${questionsPageBasePath}/:id`} component={QuestionDetails} />
  </Switch>
);
