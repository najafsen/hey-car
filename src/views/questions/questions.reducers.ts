import { createReducer, isAnyOf } from "@reduxjs/toolkit";
import {
  getQuestionByIdAction,
  getQuestionsListAction,
  voteQuestionChoiceAction,
} from "./questions.actions";
import { Question } from "./questions.types";

export interface QuestionsState {
  list: Question[];
  isNetworkRequestFailed: boolean;
}

const initialState: QuestionsState = {
  list: [],
  isNetworkRequestFailed: false,
};

export const questionsReducer = createReducer(initialState, (builder) => {
  builder.addCase(getQuestionsListAction.fulfilled, (state, action) => {
    state.list = action.payload.questions;
  });

  builder.addCase(getQuestionByIdAction.fulfilled, (state, action) => {
    const questionIndex = state.list.findIndex(
      (question) => question.url === action.payload.question.url
    );

    if (questionIndex < 0) {
      state.list.push(action.payload.question);
    } else {
      state.list[questionIndex] = action.payload.question;
    }
  });

  builder.addCase(voteQuestionChoiceAction.fulfilled, (state, action) => {
    const questionIndex = state.list.findIndex(
      (question) => question.url === `/questions/${action.payload.questionId}`
    );

    if (questionIndex < 0) {
      return;
    }

    const choiceIndex = state.list[questionIndex].choices.findIndex(
      (choice) => choice.url === action.payload.choice.url
    );

    if (choiceIndex < 0) {
      return;
    }

    state.list[questionIndex].choices[choiceIndex] = action.payload.choice;
  });

  builder.addMatcher(
    isAnyOf(
      getQuestionsListAction.rejected,
      getQuestionByIdAction.rejected,
      voteQuestionChoiceAction.rejected
    ),
    (state, action) => {
      state.isNetworkRequestFailed = true;
    }
  );
});
