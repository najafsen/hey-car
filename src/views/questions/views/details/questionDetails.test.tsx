import * as reduxHooks from "../../../../hooks/redux";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Question } from "../../questions.types";
import { extractIdsFromChoiceUrl, QuestionDetails } from "./questionDetails";
import * as questionActions from "../../questions.actions";
import { questionsFixture } from "../../__fixtures__/questions";

const match = { params: { id: "56" } };
const question: Question = questionsFixture[0];

const dispatch = jest.fn();

describe("QuestionDetails", () => {
  beforeEach(() => {
    jest.spyOn(reduxHooks, "useAppDispatch").mockImplementation(() => dispatch);
  });

  it("should render correctly", () => {
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(question);
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(false);
    const { asFragment } = render(
      <QuestionDetails match={match as any} {...({} as any)} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render network error", () => {
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(null);
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(true);
    const { getByText } = render(
      <QuestionDetails match={match as any} {...({} as any)} />
    );

    expect(
      getByText(
        "Polls app need to access internet in order to work. Check your network connection."
      )
    ).toBeInTheDocument();
  });

  it("should render null when there is no question", () => {
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(null);
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(false);
    const { container } = render(
      <QuestionDetails match={match as any} {...({} as any)} />
    );

    expect(container).toMatchInlineSnapshot(`<div />`);
  });

  it("should render header and content", () => {
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(question);
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(false);
    const { getByText, getAllByRole } = render(
      <QuestionDetails match={match as any} {...({} as any)} />
    );

    expect(getByText("Questions Detail")).toBeInTheDocument(); // top header
    expect(getByText("Question: Favourite language?")).toBeInTheDocument(); // sub header
    expect(getAllByRole("row")).toHaveLength(4);
  });

  it("should dispatch vote action", () => {
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(question);
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(false);
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(question);
    jest.spyOn(reduxHooks, "useAppSelector").mockReturnValueOnce(false);

    jest
      .spyOn(questionActions, "voteQuestionChoiceAction")
      .mockImplementation((p) => JSON.stringify(p) as any);

    const { getAllByRole, getByText } = render(
      <QuestionDetails match={match as any} {...({} as any)} />
    );

    const secondRadio = getAllByRole("radio")[1];
    userEvent.click(secondRadio);

    const button = getByText("Submit your vote!");

    userEvent.click(button);

    expect(dispatch).toHaveBeenCalledWith(
      questionActions.voteQuestionChoiceAction(
        extractIdsFromChoiceUrl(question.choices[1].url)
      )
    );
  });

  describe("extractIdsFromChoiceUrl", () => {
    it("should return correct question and choice id", () => {
      expect(extractIdsFromChoiceUrl(question.choices[0].url)).toEqual({
        questionId: "12",
        choiceId: "81",
      });
    });
  });
});
