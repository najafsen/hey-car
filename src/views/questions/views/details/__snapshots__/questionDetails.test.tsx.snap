// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionDetails should render correctly 1`] = `
<DocumentFragment>
  <header>
    <h1
      class="sc-hKFxyN fbZLAc"
    >
      Questions Detail
    </h1>
  </header>
  <article>
    <h2>
      Question: Favourite language?
    </h2>
    <div
      class="sc-bdnxRM fPfiLM"
      role="table"
    >
      <div
        class="sc-gtsrHT fPjFeZ"
        role="row"
      >
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          Deutsch
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          2
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          <div
            class="sc-jSFjdj emKhFE"
            role="progressbar"
          >
            <div
              class="sc-gKAaRy blatPh"
              width="66.66666666666666"
            >
              66.67 %
            </div>
          </div>
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          <label>
            <input
              type="radio"
            />
            Vote
          </label>
        </div>
      </div>
      <div
        class="sc-gtsrHT fPjFeZ"
        role="row"
      >
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          English
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          1
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          <div
            class="sc-jSFjdj emKhFE"
            role="progressbar"
          >
            <div
              class="sc-gKAaRy hLOGOR"
              width="33.33333333333333"
            >
              33.33 %
            </div>
          </div>
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          <label>
            <input
              type="radio"
            />
            Vote
          </label>
        </div>
      </div>
      <div
        class="sc-gtsrHT fPjFeZ"
        role="row"
      >
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          Czech
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          0
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          <div
            class="sc-jSFjdj emKhFE"
            role="progressbar"
          >
            <div
              class="sc-gKAaRy iMEojH"
              width="0"
            >
              0.00 %
            </div>
          </div>
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          <label>
            <input
              type="radio"
            />
            Vote
          </label>
        </div>
      </div>
      <div
        class="sc-gtsrHT fPjFeZ"
        role="row"
      >
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          Spanish
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          0
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          <div
            class="sc-jSFjdj emKhFE"
            role="progressbar"
          >
            <div
              class="sc-gKAaRy iMEojH"
              width="0"
            >
              0.00 %
            </div>
          </div>
        </div>
        <div
          class="sc-dlnjwi cJyYRj"
          role="cell"
        >
          <label>
            <input
              type="radio"
            />
            Vote
          </label>
        </div>
      </div>
    </div>
  </article>
  <div
    class="sc-eCApnc koQIzQ"
  >
    <button
      aria-label="Submit your vote!"
      disabled=""
      type="submit"
    >
      Submit your vote!
    </button>
  </div>
</DocumentFragment>
`;
